packages/components/src/lib/flamegraph.js

```diff
@@ -0,0 +1,66 @@
+const PRECISION = 10e6;
+
+// TODO aggregate with codeObject.fqid
+
+const digestEvent = (event) => ({
+  name: String(event.toString()),
+  value: Number(event.elapsedTime),
+  children: event.children.map(digestEvent),
+});
+
+const validateValue = (name, total, value) => {
+  if (value < 0) {
+    console.warn(`found negative value ${value} for ${name}`);
+    return total;
+  } else if (total > value) {
+    console.warn(
+      `had to increase value of ${name} from ${value} to ${total} to match children's total`
+    );
+    return total;
+  } else {
+    return value;
+  }
+};
+
+const accumulateValue = (sum, { value }) => sum + value;
+
+const validateNode = ({ name, value, children }) => {
```

Comment:
Maybe this is nit-picky, but a comment would be useful here for clarity. This isn't a "validator" in the sense that it's validating some set of criteria on the node, it's literally making the node valid by mutating/reducing the value.

Reply:
I rename it to `toCompatibleValue`.

Reply:
0a8ef4a94c6a44a54cf0481ae781669948b2f56e